plugins/pipelines/app/models/concerns/samson_pipelines/stage_concern.rb

```diff
@@ -1,15 +1,26 @@
 module SamsonPipelines::StageConcern
-  # Return true if any stages in the pipeline are marked production
-  def production?
-    super || next_stages.any?(&:production?)
+
+  def deploy_requires_approval?
+    super || all_next_stages.any?(&:deploy_requires_approval?)
   end
 
   def next_stages
-    Stage.find(next_stage_ids)
+    @next_stages ||= Stage.find(next_stage_ids)
+  end
+
+  def recursive_next_stage_ids
+    return [] if next_stage_ids.empty?
+    next_stages.collect(&:id) + next_stages.flat_map(&:recursive_next_stage_ids)
```

Comment:
FYI most code I see uses `map` instead of `collect` and it is also nicely matches to flat_map + it's shorter :)